### PR TITLE
Onlykey agent merge existing functionality

### PR DESCRIPTION
**Pull request by onlykey**
@romanz Per discussion here - https://github.com/romanz/trezor-agent/issues/303

I removed everything related to RSA and removed the replacement of ed25519 library with pynacl. Only the minimal things needed to add OnlyKey support for all existing functionality is here.

Some things to take note of
- OnlyKey has on device PIN entry so the GPG config file created for OnlyKey and the UI options are not used
- Everything that is OnlyKey specific has been wrapped in - if device_type.package_name() == 'onlykey-agent': to ensure there are no issues 
- The only additional feature here is that -sk and -dk flags are available for OnlyKey, this was also wrapped with if device_type.package_name() == 'onlykey-agent' so these options do not affect the other agents in any way
- Testing was conducted with OnlyKey supporting the derived keys for ed25519, curve25519 (ecdh), and nist256p1
- Testing was conducted on Trezor to double check everything
- Version number bumped to 0.14.2

The things that are working that I would like to do a future PR for are
- swap out ed25519 library with pynacl. It looks like ed25519 isn't maintained anymore. Also it requires build tools on windows so if you replace with pynacl you can then just pip install trezor_agent on windows without building tool requirement. I have already tested this out it's an easy drop in replacement.
- Key import - OnlyKey supports stored/resident keys for ECC and RSA. This has been tested and works well to add an -i flag followed by a PGP public key. The public key is imported to GPG and works well for sign/decrypt the same as the derived keys. For example, this was tested with a curve25519/ed25519 key created in protonmail and allows using this key with GPG+OnlyKey.
- RSA SSH - OnlyKey supports RSA keys and this works good for SSH
- RSA GPG - This does not work yet, something with the encoding of the public key I still need to figure out. 
- Add Windows support - Still need to look into this more but I think its possible. Maybe if not with GPG then with pgpy, a pure python PGP implementation.

Thanks again this is a great project!

### CODE REVIEW

**Comment by romanz** [comment]
I have cherry-picked `pynacl` support into `master` using https://github.com/romanz/trezor-agent/commit/e1bbdb4bccb9c81a34123cc89fbb6ef2750ab33b - seems to work well, many thanks!

**Comment by romanz** [comment]
Could you please squash the commits and rebase over latest `master`?

**Comment by romanz** [comment]
It still seems to fail on `pycodestyle` - you can reproduce it locally using:
```
$ tox
GLOB sdist-make: /media/roman/3TB/roman/Code/trezor-agent/setup.py
py3 inst-nodeps: /media/roman/3TB/roman/Code/trezor-agent/.tox/.tmp/package/1/libagent-0.14.2.zip
py3 installed: appdirs==1.4.3,astroid==2.4.2,attrs==19.3.0,backports.shutil-which==3.5.2,CacheControl==0.12.6,certifi==2019.11.28,chardet==3.0.4,colorama==0.4.3,ConfigArgParse==1.2.3,contextlib2==0.6.0,coverage==5.2.1,distlib==0.3.0,distro==1.4.0,docutils==0.16,ecdsa==0.15,ed25519==1.5,html5lib==1.0.1,idna==2.8,iniconfig==1.0.1,ipaddr==2.2.0,isort==5.3.0,lazy-object-proxy==1.4.3,libagent==0.14.2,lockfile==0.12.2,mccabe==0.6.1,mnemonic==0.19,mock==4.0.2,more-itertools==8.4.0,msgpack==0.6.2,packaging==20.3,pep517==0.8.2,pluggy==0.13.1,progress==1.5,py==1.9.0,pycodestyle==2.6.0,pydocstyle==5.0.2,pylint==2.5.3,PyMsgBox==1.0.8,pyparsing==2.4.6,pytest==6.0.1,python-daemon==2.2.4,pytoml==0.1.21,requests==2.22.0,retrying==1.3.3,semver==2.10.2,six==1.14.0,snowballstemmer==2.0.0,toml==0.10.1,Unidecode==1.1.1,urllib3==1.25.8,webencodings==0.5.1,wrapt==1.12.1
py3 run-test-pre: PYTHONHASHSEED='547349516'
py3 run-test: commands[0] | pycodestyle libagent
libagent/device/onlykey.py:121:86: W291 trailing whitespace
libagent/device/onlykey.py:133:24: E225 missing whitespace around operator
libagent/device/onlykey.py:142:28: E225 missing whitespace around operator
libagent/device/onlykey.py:144:24: E225 missing whitespace around operator
libagent/device/onlykey.py:146:33: E251 unexpected spaces around keyword / parameter equals
libagent/device/onlykey.py:146:35: E251 unexpected spaces around keyword / parameter equals
libagent/device/onlykey.py:146:67: W291 trailing whitespace
libagent/device/onlykey.py:147:40: E127 continuation line over-indented for visual indent
libagent/device/onlykey.py:147:47: E251 unexpected spaces around keyword / parameter equals
libagent/device/onlykey.py:147:49: E251 unexpected spaces around keyword / parameter equals
libagent/device/onlykey.py:153:62: E251 unexpected spaces around keyword / parameter equals
libagent/device/onlykey.py:153:64: E251 unexpected spaces around keyword / parameter equals
libagent/device/onlykey.py:196:101: E501 line too long (105 > 100 characters)
libagent/device/onlykey.py:196:106: W291 trailing whitespace
libagent/device/onlykey.py:197:48: E111 indentation is not a multiple of four
libagent/device/onlykey.py:197:48: E113 unexpected indentation
libagent/device/onlykey.py:197:101: E501 line too long (113 > 100 characters)
libagent/device/onlykey.py:198:101: E501 line too long (194 > 100 characters)
libagent/device/onlykey.py:200:101: E501 line too long (105 > 100 characters)
libagent/device/onlykey.py:201:48: E111 indentation is not a multiple of four
libagent/device/onlykey.py:201:48: E113 unexpected indentation
libagent/device/onlykey.py:201:101: E501 line too long (113 > 100 characters)
libagent/device/onlykey.py:208:50: E251 unexpected spaces around keyword / parameter equals
libagent/device/onlykey.py:208:52: E251 unexpected spaces around keyword / parameter equals
libagent/device/onlykey.py:212:29: W291 trailing whitespace
libagent/device/onlykey.py:267:14: E211 whitespace before '('
libagent/device/onlykey.py:268:14: E211 whitespace before '('
libagent/device/onlykey.py:271:44: E251 unexpected spaces around keyword / parameter equals
libagent/device/onlykey.py:271:46: E251 unexpected spaces around keyword / parameter equals
libagent/device/onlykey.py:271:81: E251 unexpected spaces around keyword / parameter equals
libagent/device/onlykey.py:271:83: E251 unexpected spaces around keyword / parameter equals
libagent/device/onlykey.py:271:96: W291 trailing whitespace
libagent/device/onlykey.py:272:44: E127 continuation line over-indented for visual indent
libagent/device/onlykey.py:272:51: E251 unexpected spaces around keyword / parameter equals
libagent/device/onlykey.py:272:53: E251 unexpected spaces around keyword / parameter equals
libagent/device/onlykey.py:290:44: E251 unexpected spaces around keyword / parameter equals
libagent/device/onlykey.py:290:46: E251 unexpected spaces around keyword / parameter equals
libagent/device/onlykey.py:290:87: W291 trailing whitespace
libagent/device/onlykey.py:291:44: E127 continuation line over-indented for visual indent
libagent/device/onlykey.py:291:51: E251 unexpected spaces around keyword / parameter equals
libagent/device/onlykey.py:291:53: E251 unexpected spaces around keyword / parameter equals
libagent/device/onlykey.py:291:66: E202 whitespace before ')'
libagent/device/onlykey.py:311:50: E251 unexpected spaces around keyword / parameter equals
libagent/device/onlykey.py:311:52: E251 unexpected spaces around keyword / parameter equals
libagent/device/onlykey.py:318:86: W291 trailing whitespace
libagent/device/onlykey.py:328:48: W291 trailing whitespace
libagent/device/onlykey.py:330:81: W291 trailing whitespace
libagent/device/onlykey.py:354:40: E251 unexpected spaces around keyword / parameter equals
libagent/device/onlykey.py:354:42: E251 unexpected spaces around keyword / parameter equals
libagent/device/onlykey.py:354:80: E251 unexpected spaces around keyword / parameter equals
libagent/device/onlykey.py:354:82: E251 unexpected spaces around keyword / parameter equals
libagent/device/onlykey.py:354:95: W291 trailing whitespace
libagent/device/onlykey.py:355:40: E127 continuation line over-indented for visual indent
libagent/device/onlykey.py:355:47: E251 unexpected spaces around keyword / parameter equals
libagent/device/onlykey.py:355:49: E251 unexpected spaces around keyword / parameter equals
libagent/device/onlykey.py:356:14: E211 whitespace before '('
libagent/device/onlykey.py:357:14: E211 whitespace before '('
libagent/device/onlykey.py:368:15: E271 multiple spaces after keyword
libagent/device/onlykey.py:390:1: E305 expected 2 blank lines after class or function definition, found 1
libagent/device/onlykey.py:400:1: E302 expected 2 blank lines, found 1
libagent/device/onlykey.py:404:28: W292 no newline at end of file
libagent/gpg/__init__.py:77:40: E127 continuation line over-indented for visual indent
libagent/gpg/__init__.py:78:40: E127 continuation line over-indented for visual indent
libagent/gpg/client.py:34:48: E128 continuation line under-indented for visual indent
ERROR: InvocationError for command /media/roman/3TB/roman/Code/trezor-agent/.tox/py3/bin/pycodestyle libagent (exited with code 1)
___________________________________________________________________________________________________________________________________ summary ____________________________________________________________________________________________________________________________________
ERROR:   py3: commands failed
```

**Comment by onlykey** [comment]
@romanz Apologies, I ran out of time the other day, trying to finish this up today

**Comment by romanz** [comment]
No worries :)

**Comment by onlykey** [comment]
@romanz I believe this is good to go now. I used tox as you suggested and fixed issues there. I also merged in romanz/master.

**Comment by romanz** [comment]
Could you please squash and rebase over `master`?

**Comment by romanz** [comment]
I have used the following commands:
```
$ git commit-tree 94c4206^{tree} -p origin/master -m "Add OnlyKey support"
0a520876a951156ffb1435759c92278b8ad1808d
$ git rst 0a520876a951156ffb1435759c92278b8ad1808d
HEAD is now at 0a52087 Add OnlyKey support
$ git diff onlykey/onlykey-agent-merge-existing-functionality
```
May I force-push https://github.com/romanz/trezor-agent/commit/f570f68f1b2373b32cc5e8b46623a2994298cde8 into this branch?

**Review by romanz — DISMISSED** [review]
Could you please fix the remaining CI issues on Travis?
https://travis-ci.org/github/romanz/trezor-agent/jobs/716161214

**Comment by onlykey** [comment]
@romanz I fixed the issues with pydocstyle and the one test that was failing "libagent/ssh/tests/test_client.py::test_ssh_agent PASSED "

I fixed everything in tox I could, it says Your code has been rated at 9.94/10 (previous run: 9.92/10, +0.02) but it still has these items that I am not sure how to fix, would these be OK to ignore?

libagent/device/onlykey.py:97:4: R0912: Too many branches (31/12) (too-many-branches)
libagent/device/onlykey.py:97:4: R0915: Too many statements (81/50) (too-many-statements)
libagent/device/onlykey.py:207:4: R0914: Too many local variables (18/15) (too-many-locals)
libagent/device/onlykey.py:207:4: R0912: Too many branches (24/12) (too-many-branches)
libagent/device/onlykey.py:207:4: R0915: Too many statements (81/50) (too-many-statements)
libagent/device/onlykey.py:309:4: R0914: Too many local variables (18/15) (too-many-locals)
libagent/device/onlykey.py:309:4: R0912: Too many branches (18/12) (too-many-branches)
libagent/device/onlykey.py:309:4: R0915: Too many statements (63/50) (too-many-statements)

^ For these the function does quite a few things so there are quite a few branches, statements, etc.

libagent/device/onlykey.py:39:12: W0201: Attribute 'ok' defined outside __init__ (attribute-defined-outside-init)
libagent/device/onlykey.py:38:12: W0201: Attribute 'device_name' defined outside __init__ (attribute-defined-outside-init)
libagent/device/onlykey.py:53:8: W0201: Attribute 'dkeyslot' defined outside __init__ (attribute-defined-outside-init)
libagent/device/onlykey.py:48:8: W0201: Attribute 'skeyslot' defined outside __init__ (attribute-defined-outside-init)
libagent/device/onlykey.py:58:8: W0201: Attribute 'import_pubkey' defined outside __init__ (attribute-defined-outside-init)
libagent/device/onlykey.py:89:12: W0201: Attribute 'sighash' defined outside __init__ (attribute-defined-outside-init)
libagent/device/onlykey.py:41:12: W0201: Attribute 'okversion' defined outside __init__ (attribute-defined-outside-init)
libagent/device/onlykey.py:42:12: W0201: Attribute 'okversion' defined outside __init__ (attribute-defined-outside-init)

^ For these the attributes are not defined inside __init__ because these attributes are only for the onlykey-agent and don't apply to the other agents

**Comment by romanz** [comment]
Thanks, will take a look in the weekend.

**Comment by onlykey** [comment]
@romanz Great, let me know if you need anything.

**Comment by onlykey** [comment]
@romanz Just wanted to check if you were still able to take a look this weekend

**Comment by romanz** [comment]
Sorry, I didn't had the time... :(

**Comment by onlykey** [comment]
@romanz Just wanted to check in to see if you know when you will be able to take a look.

**Comment by romanz** [comment]
Sorry for the unavailability - hope that I'll have more time in the next week...

**Comment by onlykey** [comment]
@romanz I think everything is ready to merge here, just need to ignore R0912, R0914, R0915, and W0201 from CI.

**Comment by romanz** [comment]
Sorry for the very delayed response... 

I have squashed the commits and rebased over `master` - please take a look at:
https://github.com/romanz/trezor-agent/compare/dev/onlykey

I also removed a few onlykey-specific changes (related to `skey` and `dkey` handling) - I prefer to add them in a separate PR.

Please review - and if it works with the device, please force-push it into this branch, so we can discuss a few remaining issues.

**Comment by onlykey** [comment]
@romanz Great, I will need to retest with OnlyKey hardware to ensure no issues. Next week I should have time to finish that testing/review and will let you know. Thanks!

**Comment by onlykey** [comment]
@romanz Removing the skey and dkey options required a very small change so I just did a patch in a new PR here - https://github.com/romanz/trezor-agent/pull/336

Everything is tested and working now, ready to merge.

**Comment by romanz** [comment]
https://github.com/romanz/trezor-agent/pull/337#issuecomment-698292336

**Comment by romanz** [comment]
@onlykey Could you please review https://github.com/romanz/trezor-agent/pull/337?

**Comment by romanz** [comment]
Merged, thanks!

File: libagent/gpg/__init__.py [hunk 1]
```diff
@@ -226,13 +238,18 @@ def run_agent(device_type):
...
                    help='Use stdin/stdout for communication with GPG.')
-
-    p.add_argument('--pin-entry-binary', type=str, default='pinentry',
-                   help='Path to PIN entry UI helper.')
-    p.add_argument('--passphrase-entry-binary', type=str, default='pinentry',
-                   help='Path to passphrase entry UI helper.')
-    p.add_argument('--cache-expiry-seconds', type=float, default=float('inf'),
-                   help='Expire passphrase from cache after this duration.')
+    if device_type.package_name() == 'onlykey-agent':
+        p.add_argument('-sk', '--skey', type=int, metavar='SIGN_KEY',
```
Comment (kalbasit) [inline libagent/gpg/__init__.py:242]: @onlykey @romanz the -sk feature did not make in #337. Was that on purpose?

Comment (romanz) [reply libagent/gpg/__init__.py:242]: Not on purpose, I just wanted to simplify #337.
PRs are welcome :)

**Pull request closed without merging**